com.oracle.truffle.r.runtime/src/com/oracle/truffle/r/runtime/data/RAltIntVectorData.java

```diff
@@ -0,0 +1,286 @@
+     *   Duplicate. The behavior may change, so the implementation is not optimized.
+     */
+    @ExportMessage
+    public int[] getIntDataCopy(@Cached AltrepDuplicateNode duplicateNode,
+                                @Cached AltrepRFFI.AltIntDataptrNode dataptrNode,
+                                @Cached AltrepLengthNode lengthNode) {
+        // TODO: deep?
+        Object duplicatedObject = duplicateNode.execute(getOwner(), false);
+        assert duplicatedObject instanceof RAltIntVectorData || duplicatedObject instanceof RIntArrayVectorData;
+        if (duplicatedObject instanceof RAltIntVectorData) {
+            return ((RAltIntVectorData) duplicatedObject).getDataCopy(dataptrNode, lengthNode);
+        } else if (duplicatedObject instanceof RIntArrayVectorData) {
+            return ((RIntArrayVectorData) duplicatedObject).getIntDataCopy();
+        } else {
+            throw RInternalError.shouldNotReachHere(
+                    "RAltIntVectorData.getIntDataCopy: Unexpected returned object = " + duplicatedObject.toString()
+            );
+        }
+    }
+
+    @ExportMessage
+    public int[] getReadonlyIntData(@Cached AltrepRFFI.AltIntDataptrNode dataptrNode,
+                                    @Cached AltrepLengthNode lengthNode) {
+        return getDataCopy(dataptrNode, lengthNode);
+    }
+
+    private int[] getDataCopy(AltrepRFFI.AltIntDataptrNode dataptrNode, AltrepLengthNode lengthNode) {
+        long dataptrAddr = dataptrNode.execute(getOwner(), false);
+        int length = (int) lengthNode.execute(getOwner());
+        int[] dataCopy = new int[length];
+        NativeMemory.copyMemory(dataptrAddr, dataCopy, ElementType.INT, length);
+        return dataCopy;
+    }
+
+    @ExportMessage
+    public SeqIterator iterator(
+            @Shared("SeqItLoopProfile") @Cached("createCountingProfile()") LoopConditionProfile loopProfile,
+            @Cached AltrepLengthNode lengthnode) {
+        SeqIterator it = new SeqIterator(this, (int) lengthnode.execute(getOwner()));
+        it.initLoopConditionProfile(loopProfile);
+        return it;
+    }
+
+    @ExportMessage
+    public RandomAccessIterator randomAccessIterator() {
+        return new RandomAccessIterator(this);
+    }
+
+    @ExportMessage
+    public boolean next(SeqIterator it, boolean withWrap,
+                        @Shared("SeqItLoopProfile") @Cached("createCountingProfile()") LoopConditionProfile loopProfile) {
+        return it.next(loopProfile, withWrap);
+    }
+
+    @ExportMessage(limit = "1")
+    public int getIntAt(int index,
+                        @Cached(value = "create()") AltrepGetIntAtNode getIntAtNode) {
+        return (int) getIntAtNode.execute(getOwner(), index);
+    }
+
+    @ExportMessage(limit = "1")
+    public int getNextInt(SeqIterator it,
+                          @Cached(value = "create()") AltrepGetIntAtNode getIntAtNode) {
+        return (int) getIntAtNode.execute(getOwner(), it.getIndex());
```

Comment:
I was thinking we could take advantage of the API that gives you contiguous chunks of memory, you can store some data structure into the iterator that would help you with advancing the iterator and requesting a new chunk when you're done with the current one.

Reply:
In the meeting, we should discuss how these iterators are supposed to work together with Truffle compiler